### PR TITLE
Add Vector and Matrix compare, equal and equals with epsilon parameter

### PR DESCRIPTION
**Pull request by ThomasGorisse**
Discussion #47

### CODE REVIEW

File: src/commonMain/kotlin/dev/romainguy/kotlin/math/Matrix.kt [hunk 1]
```diff
@@ -385,6 +385,117 @@ data class Mat4(
     }
 }
 
+inline fun compare(a: Mat2, b: Float, eps: Float? = null): Mat2 {
```
Comment (romainguy) [inline src/commonMain/kotlin/dev/romainguy/kotlin/math/Matrix.kt:388]: You could just write 
```
) = Mat2(
   ...
)
```
instead

Comment (romainguy) [reply src/commonMain/kotlin/dev/romainguy/kotlin/math/Matrix.kt:388]: The compare methods should be called compareTo() to match Java/Kotlin

File: src/commonMain/kotlin/dev/romainguy/kotlin/math/Quaternion.kt [hunk 2]
```diff
@@ -294,6 +294,36 @@ inline operator fun Float.minus(q: Quaternion) = Quaternion(this - q.x, this - q
 inline operator fun Float.times(q: Quaternion) = Quaternion(this * q.x, this * q.y, this * q.z, this * q.w)
 inline operator fun Float.div(q: Quaternion) = Quaternion(this / q.x, this / q.y, this / q.z, this / q.w)
 
+inline fun compare(a: Quaternion, b: Float, eps: Float? = null) = compare(normalize(a).xyzw, b, eps)
```
Comment (romainguy) [inline src/commonMain/kotlin/dev/romainguy/kotlin/math/Quaternion.kt:297]: Why are you normalizing before the tests?

File: src/commonMain/kotlin/dev/romainguy/kotlin/math/Vector.kt [hunk 3]
```diff
@@ -567,6 +567,23 @@ inline operator fun Float.minus(v: Float2) = Float2(this - v.x, this - v.y)
 inline operator fun Float.times(v: Float2) = Float2(this * v.x, this * v.y)
 inline operator fun Float.div(v: Float2) = Float2(this / v.x, this / v.y)
 
+inline fun compare(a: Float, b: Float, eps: Float? = null): Float = when {
```
Comment (romainguy) [inline src/commonMain/kotlin/dev/romainguy/kotlin/math/Vector.kt:570]: Use https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-float/compare-to.html instead. Or at least write a version that supports NaN properly:

```
        if (a > b) return 1
        if (a < b) return -1

        return a.toBits().compareTo(b.toBits())
```

File: src/commonMain/kotlin/dev/romainguy/kotlin/math/Vector.kt [hunk 4]
```diff
@@ -567,6 +567,23 @@ inline operator fun Float.minus(v: Float2) = Float2(this - v.x, this - v.y)
...
+    equals(a, b, eps) -> 0.0f
+    a < b -> -1.0f
+    else -> 1.0f
+}
+inline fun lessThan(a: Float, b: Float, eps: Float? = null) = compare(a, b, eps) < 0
+inline fun lessThanEqual(a: Float, b: Float, eps: Float? = null) = compare(a, b, eps) <= 0
+inline fun greaterThan(a: Float, b: Float, eps: Float? = null) = compare(a, b, eps) > 0
+inline fun greaterThanEqual(a: Float, b: Float, eps: Float? = null) = compare(a, b, eps) >= 0
+inline fun equals(a: Float, b: Float, eps: Float?): Boolean {
```
Comment (romainguy) [inline src/commonMain/kotlin/dev/romainguy/kotlin/math/Vector.kt:579]: Don't make eps nullable, it causes boxing. Set it to 0.0f by default.

**Review by romainguy — CHANGES_REQUESTED** [review]

**Comment by ThomasGorisse** [comment]
"J'ai enfin revu ma copie!"

Sorry for the late. I was in the middle of an onboarding and didn't had time for it before.

File: src/commonMain/kotlin/dev/romainguy/kotlin/math/Vector.kt [hunk 5]
```diff
@@ -535,6 +561,12 @@ data class Float4(
     inline operator fun minus(v: Float) = Float4(x - v, y - v, z - v, w - v)
     inline operator fun times(v: Float) = Float4(x * v, y * v, z * v, w * v)
     inline operator fun div(v: Float) = Float4(x / v, y / v, z / v, w / v)
+    inline fun compareTo(v: Float, delta: Float = 0.0f) = Float4(
+        xyz.compareTo(v, delta),
```
Comment (romainguy) [inline src/commonMain/kotlin/dev/romainguy/kotlin/math/Vector.kt:565]: See below, you are creating too many objects for no reason

File: src/commonMain/kotlin/dev/romainguy/kotlin/math/Quaternion.kt [hunk 6]
```diff
@@ -294,6 +299,32 @@ inline operator fun Float.minus(q: Quaternion) = Quaternion(this - q.x, this - q
 inline operator fun Float.times(q: Quaternion) = Quaternion(this * q.x, this * q.y, this * q.z, this * q.w)
 inline operator fun Float.div(q: Quaternion) = Quaternion(this / q.x, this / q.y, this / q.z, this / q.w)
 
+inline fun lessThan(a: Quaternion, b: Float) = lessThan(a.xyzw, b)
```
Comment (romainguy) [inline src/commonMain/kotlin/dev/romainguy/kotlin/math/Quaternion.kt:302]: Each .xyzw creates a new vector

File: src/commonMain/kotlin/dev/romainguy/kotlin/math/Quaternion.kt [hunk 7]
```diff
@@ -294,6 +299,32 @@ inline operator fun Float.minus(q: Quaternion) = Quaternion(this - q.x, this - q
...
+inline fun greaterThan(a: Quaternion, b: Float) = greaterThan(a.xyzw, b)
+inline fun greaterThan(a: Quaternion, b: Quaternion) = greaterThan(a.xyzw, b.xyzw)
+inline fun greaterThanEqual(a: Quaternion, b: Float) = greaterThanEqual(a.xyzw, b)
+inline fun greaterThanEqual(a: Quaternion, b: Quaternion) = greaterThanEqual(a.xyzw, b.xyzw)
+inline fun equal(a: Quaternion, b: Float) = equal(a.xyzw, b)
+inline fun equal(a: Quaternion, b: Quaternion) = equal(a.xyzw, b.xyzw)
+inline fun notEqual(a: Quaternion, b: Float) = notEqual(a.xyzw, b)
+inline fun notEqual(a: Quaternion, b: Quaternion) = notEqual(a.xyzw, b.xyzw)
+
+inline infix fun Quaternion.lt(b: Float) = xyzw.lt(b)
```
Comment (romainguy) [inline src/commonMain/kotlin/dev/romainguy/kotlin/math/Quaternion.kt:315]: Same here: Each .xyzw creates a new vector

**Pull request merged**